local/blockchain.go

```diff
@@ -769,41 +847,144 @@ func importPChainFromXChain(ctx context.Context, w *wallet, owner *secp256k1fx.O
 	return err
 }
 
+func (ln *localNetwork) addPermissionlessValidators(
+	ctx context.Context,
+	validatorSpecs []network.PermissionlessValidatorSpec,
+) ([]ids.ID, error) {
+	ln.log.Info("adding permissionless validator tx")
+	validatorSpecIDs := make([]ids.ID, len(validatorSpecs))
+	clientURI, err := ln.getClientURI()
+	if err != nil {
+		return nil, err
+	}
+	platformCli := platformvm.NewClient(clientURI)
+	// wallet needs txs for all previously created subnets
+	var preloadTXs []ids.ID
+	for _, validatorSpec := range validatorSpecs {
+		if validatorSpec.SubnetID == nil {
+			return nil, errors.New("permissionless validator spec has no subnet ID")
+		} else {
+			subnetID, err := ids.FromString(*validatorSpec.SubnetID)
+			if err != nil {
+				return nil, err
+			}
+			preloadTXs = append(preloadTXs, subnetID)
+		}
+	}
+	w, err := newWallet(ctx, clientURI, preloadTXs)
+	if err != nil {
+		return nil, err
+	}
+	owner := &secp256k1fx.OutputOwners{
+		Threshold: 1,
+		Addrs: []ids.ShortID{
+			w.addr,
+		},
+	}
+	// create new nodes
+	for _, validatorSpec := range validatorSpecs {
+		_, ok := ln.nodes[validatorSpec.NodeName]
+		if !ok {
+			ln.log.Info(logging.Green.Wrap(fmt.Sprintf("adding new participant %s", validatorSpec.NodeName)))
+			if _, err := ln.addNode(node.Config{Name: validatorSpec.NodeName}); err != nil {
+				return nil, err
+			}
+		}
+	}
+	if err := ln.healthy(ctx); err != nil {
+		return nil, err
+	}
+
+	// just ensure all nodes are primary validators (so can be subnet validators)
+	if err := ln.addPrimaryValidators(ctx, platformCli, w); err != nil {
+		return nil, err
+	}
+
+	// wait for nodes to be primary validators before trying to add them as subnet ones
+	if err = ln.waitPrimaryValidators(ctx, platformCli); err != nil {
+		return nil, err
+	}
+
+	for i, validatorSpec := range validatorSpecs {
+		ln.log.Info(logging.Green.Wrap("adding permissionless validator"), zap.String("node ", validatorSpec.NodeName))
+		cctx, cancel := createDefaultCtx(ctx)
+		validatorNodeID := ln.nodes[validatorSpec.NodeName].nodeID
+		subnetID, err := ids.FromString(*validatorSpec.SubnetID)
+		if err != nil {
+			return nil, err
+		}
+		assetID, err := ids.FromString(validatorSpec.AssetID)
+		if err != nil {
+			return nil, err
+		}
+		txID, err := w.pWallet.IssueAddPermissionlessValidatorTx(
+			&txs.SubnetValidator{
+				Validator: txs.Validator{
+					NodeID: validatorNodeID,
+					Start:  uint64(validatorSpec.StartTime.Unix()),
```

Comment:
can we have defaults for start/end non given params? like start to be 20s in future, end to be end of primary period if already primary val, wght to be equal to default weight used in the other add validator call.
Can